### PR TITLE
feat: Validate if a BuildItem is `final` or `abstract`

### LINKED ISSUE #1120: Validate if a BuildItem is `final` or `abstract`
BuildItems consumed or produced in BuildSteps must be final or abstract. It would be nice if the IDE could warn about that when creating a new BuildItem subclass.

Eg. 
`public abstract class AbstractDevUIBuildItem extends MultiBuildItem {...}`
`public final class QuteTemplateBuildItem extends AbstractDevUIBuildItem {...}`

### PR DESCRIPTION
**Pull request by fbricon**
Validate if a BuildItem is `final` or `abstract`

Screenshot 2023-09-15 at 11 11 50

cc @gastaldi 

Fixes #1120 

I still need to write tests

### CODE REVIEW

**Comment by angelozerr** [comment]
Do you think you could provide code action to insert abstract / final keyword, or perhaps in an another PR?

File: src/main/java/com/redhat/devtools/intellij/quarkus/psi/internal/validators/QuarkusBuildItemDiagnosticsParticipant.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                createDiagnosticMessage(psiClass, context.getDocumentFormat()),
+                range, "quarkus",
+                QuarkusBuildItemErrorCode.InvalidModifierBuildItem,
+                DiagnosticSeverity.Error
+        );
+        diagnostics.add(d);
+    }
+
+    private static String createDiagnosticMessage(PsiClass classType, DocumentFormat documentFormat) {
+        StringBuilder message = new StringBuilder("BuildItem class ");
```
Comment (angelozerr) [inline src/main/java/com/redhat/devtools/intellij/quarkus/psi/internal/validators/QuarkusBuildItemDiagnosticsParticipant.java:89]: Use a message format (by declaring the key message as static)

File: src/main/java/com/redhat/devtools/intellij/quarkus/psi/internal/validators/QuarkusBuildItemDiagnosticsParticipant.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import com.redhat.devtools.intellij.quarkus.psi.internal.builditems.QuarkusBuildItemErrorCode;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.DiagnosticSeverity;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4mp.commons.DocumentFormat;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class QuarkusBuildItemDiagnosticsParticipant implements IJavaDiagnosticsParticipant {
```
Comment (angelozerr) [inline src/main/java/com/redhat/devtools/intellij/quarkus/psi/internal/validators/QuarkusBuildItemDiagnosticsParticipant.java:37]: Add Javadoc which explains the goal of this validator

Comment (fbricon) [reply src/main/java/com/redhat/devtools/intellij/quarkus/psi/internal/validators/QuarkusBuildItemDiagnosticsParticipant.java:43]: done

**Comment by fbricon** [comment]
Added some tests

**Comment by fbricon** [comment]
> Do you think you could provide code action to insert abstract / final keyword, or perhaps in an another PR?

maybe another PR

File: src/main/java/com/redhat/devtools/intellij/quarkus/psi/internal/validators/QuarkusBuildItemDiagnosticsParticipant.java (hunk 1 above)
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/quarkus/psi/internal/validators/QuarkusBuildItemDiagnosticsParticipant.java]: @fbricon I agree that using à prooérties file is cleaner but we need to think about translation in JDT. Using a static string key for message will avoid translating this code

**Comment by angelozerr** [comment]
> maybe another PR

OK please create an issue to avoid forgetting to implement this code action

File: src/main/java/com/redhat/devtools/intellij/quarkus/psi/internal/validators/QuarkusBuildItemDiagnosticsParticipant.java (hunk 1 above)
Comment (fbricon) [reply src/main/java/com/redhat/devtools/intellij/quarkus/psi/internal/validators/QuarkusBuildItemDiagnosticsParticipant.java]: done

File: src/test/java/com/redhat/devtools/intellij/quarkus/validation/BuildItemDiagnosticsTest.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import static com.redhat.devtools.intellij.lsp4mp4ij.psi.core.MicroProfileForJavaAssert.*;
+
+public class BuildItemDiagnosticsTest extends MavenModuleImportingTestCase {
+
+    public void testBuildItemClassifier() throws Exception {
+        Module module = createMavenModule(new File("projects/quarkus/projects/maven/quarkus-builditems"));
+
+        { // Bad BuildItem shows error
+            MicroProfileJavaDiagnosticsParams diagnosticsParams = new MicroProfileJavaDiagnosticsParams();
+            VirtualFile javaFile = LocalFileSystem.getInstance().refreshAndFindFileByPath(ModuleUtilCore.getModuleDirPath(module) + "/src/main/java/org/acme/builditems/BadBuildItem.java");
```
Comment (angelozerr) [inline src/test/java/com/redhat/devtools/intellij/quarkus/validation/BuildItemDiagnosticsTest.java:40]: Please do like my current PR https://github.com/redhat-developer/intellij-quarkus/pull/1170/files#diff-48a8e83e5e1fffeb2ce83f5cd096907a8053518a68d2933826929055f4ac3671R97

`String uri = LSPIJUtils.toUri(module).resolve("src/main/java/org/acme/builditems/BadBuildItem.java").toASCIIString();`

File: src/test/java/com/redhat/devtools/intellij/quarkus/validation/BuildItemDiagnosticsTest.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                    "BuildItem class `org.acme.builditems.BadBuildItem` must either be declared final or abstract",
+                    DiagnosticSeverity.Error, QuarkusConstants.QUARKUS_DIAGNOSTIC_SOURCE,
+                    QuarkusBuildItemErrorCode.InvalidModifierBuildItem);
+            assertJavaDiagnostics(diagnosticsParams, PsiUtilsLSImpl.getInstance(myProject), d);
+        }
+
+
+        { // Good BuildItem shows no error
+            MicroProfileJavaDiagnosticsParams diagnosticsParams = new MicroProfileJavaDiagnosticsParams();
+            VirtualFile javaFile = LocalFileSystem.getInstance().refreshAndFindFileByPath(ModuleUtilCore.getModuleDirPath(module) + "/src/main/java/org/acme/builditems/GoodBuildItem.java");
```
Comment (angelozerr) [inline src/test/java/com/redhat/devtools/intellij/quarkus/validation/BuildItemDiagnosticsTest.java:55]: same here

`String uri = LSPIJUtils.toUri(module).resolve("src/main/java/org/acme/builditems/GoodBuildItem.java").toASCIIString();`

Comment (fbricon) [reply src/test/java/com/redhat/devtools/intellij/quarkus/validation/BuildItemDiagnosticsTest.java]: done

File: src/test/java/com/redhat/devtools/intellij/quarkus/validation/BuildItemDiagnosticsTest.java (hunk 3 above)
Comment (fbricon) [reply src/test/java/com/redhat/devtools/intellij/quarkus/validation/BuildItemDiagnosticsTest.java]: done

**Comment by angelozerr** [comment]
Thanks @fbricon !

**Pull request merged**